keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsStepTest.java

```diff
@@ -0,0 +1,151 @@
+package io.improbable.keanu.algorithms.mcmc;
+
+import io.improbable.keanu.DeterministicRule;
+import io.improbable.keanu.algorithms.mcmc.proposal.PriorProposalDistribution;
+import io.improbable.keanu.algorithms.mcmc.proposal.Proposal;
+import io.improbable.keanu.algorithms.mcmc.proposal.ProposalDistribution;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.UniformVertex;
+import lombok.AllArgsConstructor;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.Set;
+
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class MetropolisHastingsStepTest {
+
+    @Rule
+    public DeterministicRule deterministicRule = new DeterministicRule();
+
+    private KeanuRandom alwaysAccept;
+    private KeanuRandom alwaysReject;
+
+    @Before
+    public void setup() {
+        alwaysAccept = mock(KeanuRandom.class);
+        when(alwaysAccept.nextDouble()).thenReturn(0.0);
+
+        alwaysReject = mock(KeanuRandom.class);
+        when(alwaysReject.nextDouble()).thenReturn(1.0);
+    }
+
+    @Test
+    public void doesCalculateCorrectLogProbAfterAcceptingStep() {
+        DoubleVertex A = new GaussianVertex(0, 1);
+        A.setValue(1.0);
+        DoubleVertex B = A.times(2);
+        DoubleVertex observedB = new GaussianVertex(B, 1);
+        observedB.observe(5);
+
+        BayesianNetwork network = new BayesianNetwork(A.getConnectedGraph());
+        double logProbBeforeStep = network.getLogOfMasterP();
+
+        MetropolisHastingsStep mhStep = new MetropolisHastingsStep(
+            network.getLatentVertices(),
+            ProposalDistribution.usePrior(),
+            true,
+            KeanuRandom.getDefaultRandom()
+        );
+
+        MetropolisHastingsStep.StepResult result = mhStep.step(
+            Collections.singleton(A),
+            logProbBeforeStep
+        );
+
+        assertEquals(network.getLogOfMasterP(), result.getLogProbabilityAfterStep(), 1e-10);
+        assertTrue(result.isAccepted());
```

Comment:
How can we assert this is true? Should we be using `alwaysAccept`

Reply:
Good catch!